File: src/main/java/com/kisman/cc/util/minecraft/Packets.java
package com.kisman.cc.util.minecraft;

import io.netty.buffer.Unpooled;
import net.minecraft.client.Minecraft;
import net.minecraft.client.network.NetHandlerPlayClient;
import net.minecraft.network.INetHandler;
import net.minecraft.network.Packet;
import net.minecraft.network.PacketBuffer;
import net.minecraft.network.play.client.CPacketCustomPayload;

import java.util.Random;

/**
 * @author Cubic
 * @since 14.08.2022
 */
public class Packets {

    private static final Minecraft mc = Minecraft.getMinecraft();

    private static final NetHandlerPlayClient connection = mc.player.connection;

    private static final INetHandler handler = connection.getNetworkManager().getNetHandler();


    public static INetHandler getNetHandler(){
        return handler;
    }

    public static void processPacket(Packet packet){
        packet.processPacket(handler);
    }

    public static void send(Packet<?> packet){
        connection.sendPacket(packet);
    }

    public static void sendRandPacket(){
        Random random = new Random();
        byte[] bytes = new byte[random.nextInt(64)];
        for(int i = 0; i < bytes.length; i++)
            bytes[i] = (byte) random.nextInt(256);
        PacketBuffer buffer = new PacketBuffer(Unpooled.copiedBuffer(bytes));
        CPacketCustomPayload packet = new CPacketCustomPayload("", buffer);
        connection.sendPacket(packet);
    }

    public static void send(byte[] bytes){
        connection.sendPacket(new CPacketCustomPayload("", new PacketBuffer(Unpooled.copiedBuffer(bytes))));
    }

    public static CPacketCustomPayload newCustomPacket(byte[] bytes){
        return new CPacketCustomPayload("", new PacketBuffer(Unpooled.copiedBuffer(bytes)));
    }

    public static PacketBuffer newBuf(byte[] bytes){
        return new PacketBuffer(Unpooled.copiedBuffer(bytes));
    }

    /**
     * @param bytes may never be used again after calling this method!
     */
    public static PacketBuffer newBufUnsafe(byte[] bytes){
        return new PacketBuffer(Unpooled.wrappedBuffer(bytes));
    }
}

File: src/main/java/com/kisman/cc/util/minecraft/Packets.kt
@file:Suppress("UNCHECKED_CAST")

package com.kisman.cc.util.minecraft

import com.kisman.cc.mixin.accessors.INetworkManager
import com.kisman.cc.util.Globals.mc
import io.netty.buffer.ByteBufAllocator
import io.netty.channel.*
import io.netty.util.Attribute
import io.netty.util.AttributeKey
import io.netty.util.concurrent.EventExecutor
import net.minecraft.network.INetHandler
import net.minecraft.network.Packet
import net.minecraft.network.play.client.CPacketPlayer
import java.net.SocketAddress

/**
 * @author _kisman_
 * @since 22:28 of 10.01.2023
 */

fun sendPacketNoEvent(
    packet : Packet<*>
) {
    sendPacketNoEvent(packet, true)
}

fun sendPacketNoEvent(
    packet : Packet<*>,
    post : Boolean
) {
    if(mc.player != null) {
        (mc.player.connection.networkManager as INetworkManager).sendPacketNoEvent(packet, post)
    }
}

fun position(
    x : Double,
    y : Double,
    z : Double,
    ground : Boolean
) : CPacketPlayer.Position = CPacketPlayer.Position(
    x,
    y,
    z,
    ground
)

fun rotation(
    yaw : Float,
    pitch : Float,
    ground : Boolean
) : CPacketPlayer.Rotation = CPacketPlayer.Rotation(
    yaw,
    pitch,
    ground
)

fun positionRotation(
    x : Double,
    y : Double,
    z : Double,
    yaw : Float,
    pitch : Float,
    ground : Boolean
) : CPacketPlayer.PositionRotation = CPacketPlayer.PositionRotation(
    x,
    y,
    z,
    yaw,
    pitch,
    ground
)

fun receive(
    packet : Packet<*>
) {
    (mc.player.connection.networkManager as INetworkManager).channelRead00(null, packet)
}

fun <T : INetHandler> processPacket(
    packet : Packet<T>
) {
    packet.processPacket(mc.player.connection.networkManager.netHandler as T)
}